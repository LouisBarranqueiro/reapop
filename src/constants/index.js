export const DEFAULT_STATUS = 'default';
export const INFO_STATUS = 'info';
export const SUCCESS_STATUS = 'success';
export const WARNING_STATUS = 'warning';
export const ERROR_STATUS = 'error';

export const STATUS = {
  default: DEFAULT_STATUS,
  info: INFO_STATUS,
  success: SUCCESS_STATUS,
  warning: WARNING_STATUS,
  error: ERROR_STATUS
};

export const TOP = 't';
export const TOP_CENTER = 'tc';
export const TOP_LEFT_POSITION = 'tl';
export const TOP_RIGHT_POSITION = 'tr';
export const BOTTOM = 'b';
export const BOTTOM_CENTER = 'bc';
export const BOTTOM_LEFT_POSITION = 'bl';
export const BOTTOM_RIGHT_POSITION = 'br';

export const POSITIONS = {
  top: TOP,
  topCenter: TOP_CENTER,
  topLeft: TOP_LEFT_POSITION,
  topRight: TOP_RIGHT_POSITION,
  bottom: BOTTOM,
  bottomCenter: BOTTOM_CENTER,
  bottomLeft: BOTTOM_LEFT_POSITION,
  bottomRight: BOTTOM_RIGHT_POSITION
};

// default value for notifications
export const DEFAULT_NOTIFICATION = {
  status: DEFAULT_STATUS,
  position: TOP_RIGHT_POSITION,
  dismissible: true,
  dismissAfter: 5000,
  allowHTML: false,
  closeButton: false
};
