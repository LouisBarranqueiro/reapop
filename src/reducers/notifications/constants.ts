export enum NotificationAction {
    UpsertNotification = 'reapop/upsertNotification',
    DismissNotification = 'reapop/dismissNotification',
    DismissNotifications = 'reapop/dismissNotifications',
}
