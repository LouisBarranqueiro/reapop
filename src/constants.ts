import {NewNotification, Notification, Position} from './reducers/notifications/types'

export type NotificationConfig = {
    defaultProps: Partial<Notification>
    customizeNotification: ((notification: NewNotification) => void) | null
    generateId: ((notification: NewNotification) => string) | null
}

export const CONFIG: NotificationConfig = {
    defaultProps: {},
    customizeNotification: null,
    generateId: null,
}

export const STATUSES: {
    none: 'none'
    info: 'info'
    success: 'success'
    loading: 'loading'
    warning: 'warning'
    error: 'error'
} = Object.freeze({
    none: 'none',
    info: 'info',
    success: 'success',
    loading: 'loading',
    warning: 'warning',
    error: 'error',
})

export const POSITIONS: {
    topCenter: 'top-center'
    topLeft: 'top-left'
    topRight: 'top-right'
    bottomCenter: 'bottom-center'
    bottomLeft: 'bottom-left'
    bottomRight: 'bottom-right'
} = Object.freeze({
    topCenter: 'top-center',
    topLeft: 'top-left',
    topRight: 'top-right',
    bottomCenter: 'bottom-center',
    bottomLeft: 'bottom-left',
    bottomRight: 'bottom-right',
})

const classnamePrefix = 'reapop__'

export const classnames = {
    container: (position: Position, singleContainer: boolean) => {
        const classes = [`${classnamePrefix}container`, `${classnamePrefix}container--${position}`]
        if (singleContainer) {
            classes.push(`${classnamePrefix}container--single`)
        }
        return classes
    },
    notification: (notification: Notification) => [
        `${classnamePrefix}notification`,
        `${classnamePrefix}notification--${notification.status}`,
        `${classnamePrefix}notification--buttons-${notification.buttons.length}`,
        notification.dismissible ? `${classnamePrefix}notification--dismissible` : '',
    ],
    notificationIcon: `${classnamePrefix}notification-icon`,
    notificationImageContainer: `${classnamePrefix}notification-image-container`,
    notificationImage: `${classnamePrefix}notification-image`,
    notificationDismissIcon: `${classnamePrefix}notification-dismiss-icon`,
    notificationMeta: `${classnamePrefix}notification-meta`,
    notificationTitle: `${classnamePrefix}notification-title`,
    notificationMessage: `${classnamePrefix}notification-message`,
    notificationButtons: `${classnamePrefix}notification-buttons`,
    notificationButton: `${classnamePrefix}notification-button`,
    notificationButtonText: `${classnamePrefix}notification-button-text`,
}
