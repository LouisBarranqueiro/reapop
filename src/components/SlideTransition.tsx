import {Transition} from 'react-transition-group'
import React, {ReactNode} from 'react'
import {POSITIONS} from '../constants'
import {Notification} from '../reducers/notifications/types'

type Props = {
    notification: Notification
    children: ReactNode
    duration?: number
    [index: string]: any
}

const SlideTransition = (props: Props) => {
    const duration = props.duration || 300
    const colapseAnimationDuration = 250
    const {children, notification, ...otherProps} = props
    const transformDirection = ([POSITIONS.topCenter, POSITIONS.bottomCenter] as string[]).includes(
        notification.position
    )
        ? 'translateY'
        : 'translateX'
    const transformValue = ([POSITIONS.topCenter, POSITIONS.topLeft, POSITIONS.bottomLeft] as string[]).includes(
        notification.position
    )
        ? '-100%'
        : '100%'

    // eslint-disable-next-line no-undef
    const animationProps: KeyframeAnimationOptions = {
        fill: 'forwards',
        duration,
    }
    const onEnter = (node: HTMLElement) => {
        node.animate(
            [
                {transform: `${transformDirection}(${transformValue})`, opacity: 0},
                {transform: `${transformDirection}(0)`, opacity: 1},
            ],
            animationProps
        )
    }
    const onExit = (node: HTMLElement) => {
        const hideAnimationDuration = duration
        node.animate([{transform: `${transformDirection}(${transformValue})`, opacity: 0}], animationProps)
        setTimeout(() => {
            // `150px`: A value higher than the height a notification can have
            // to create a smooth animation for displayed notifications
            // when a notification is removed from a container.
            node.animate([{maxHeight: '150px'}, {margin: 0, maxHeight: 0}], {
                ...animationProps,
                duration: colapseAnimationDuration,
            })
        }, hideAnimationDuration)
    }

    return (
        <Transition onEnter={onEnter} onExit={onExit} timeout={duration + colapseAnimationDuration} {...otherProps}>
            {children}
        </Transition>
    )
}

export default SlideTransition
