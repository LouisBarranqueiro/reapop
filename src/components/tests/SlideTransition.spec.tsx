import React from 'react'
import {render} from '@testing-library/react'

import {POSITIONS, STATUSES} from '../../constants'

import SlideTransition from '../SlideTransition'

describe('<SlideTransition/>', () => {
    describe.each(Object.values(POSITIONS))('position: %s', (position) => {
        it('should display notification', () => {
            const baseNotification = {
                id: '1',
                position,
                status: STATUSES.none,
                buttons: [],
            }
            const {container} = render(
                <SlideTransition notification={baseNotification}>
                    <div>notification</div>
                </SlideTransition>
            )
            expect(container.innerHTML).toMatchSnapshot()
        })
    })
})
