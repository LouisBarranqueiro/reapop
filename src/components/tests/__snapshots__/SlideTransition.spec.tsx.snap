// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlideTransition/> position: bottom-center should display notification 1`] = `"<div>notification</div>"`;

exports[`<SlideTransition/> position: bottom-left should display notification 1`] = `"<div>notification</div>"`;

exports[`<SlideTransition/> position: bottom-right should display notification 1`] = `"<div>notification</div>"`;

exports[`<SlideTransition/> position: top-center should display notification 1`] = `"<div>notification</div>"`;

exports[`<SlideTransition/> position: top-left should display notification 1`] = `"<div>notification</div>"`;

exports[`<SlideTransition/> position: top-right should display notification 1`] = `"<div>notification</div>"`;
